### PR TITLE
[core] Filter data file while executing local orphan clean

### PR DESCRIPTION
**Pull request by bknbkn**
### Purpose

Local Remove Orphan Files OOM always occurred when collecting data file meta. Because  data files in tables may be large (could reach tens of millions or more). 

Relatively speaking, the number of candidate files is not large (ten thousand or hundreds of thousands).

So we can just collect data file meta in candidate files to reduce the probability of OOM occurrence.

### CODE REVIEW

File: paimon-core/src/main/java/org/apache/paimon/operation/LocalOrphanFilesClean.java [hunk 1]
```diff
@@ -101,10 +104,10 @@ public List<Path> clean() throws IOException, ExecutionException, InterruptedExc
...
 
         // delete unused files
-        Set<String> deleted = new HashSet<>(candidates.keySet());
-        deleted.removeAll(usedFiles);
-        deleted.stream().map(candidates::get).forEach(fileCleaner);
-        deleteFiles.addAll(deleted.stream().map(candidates::get).collect(Collectors.toList()));
+        candidateDeletes.removeAll(usedFiles);
+        candidateDeletes.stream().map(candidates::get).forEach(fileCleaner);
+        deleteFiles.addAll(
+                candidateDeletes.stream().map(candidates::get).collect(Collectors.toList()));
```
Comment (wwj6591812) [inline paimon-core/src/main/java/org/apache/paimon/operation/LocalOrphanFilesClean.java:111]: candidateDeletes.clear();

**Comment by wwj6591812** [comment]
I think this is a good optimization, but two questions.
1、In your production environment, you encountered OOM here? By dump heap and analyze?
2、Since you have done LocalOrphanFilesClean, can you also add this optimization to the distributed OrphanFilesClean?

**Comment by bknbkn** [comment]
I'm not sure whether this optimization can be used in a distributed situation, because that means that the candidate set needs to be broadcast to each executor. At this time, the number of data file meta in each executor is not large, and there is no risk of OOM. 

On the other hand, the process of joining condidate set and used file set actually implicitly includes the process of broadcasting and filtering.

And the analyze in production environment will be given later.

File: paimon-core/src/main/java/org/apache/paimon/operation/LocalOrphanFilesClean.java (hunk 1 above)
Comment (bknbkn) [reply paimon-core/src/main/java/org/apache/paimon/operation/LocalOrphanFilesClean.java:112]: fixed

**Comment by bknbkn** [comment]
Test in table with 7TB size:

**Before this patch, it will OOM when it collect 80 million data file metas:**
490d8adf-0f08-42f8-8453-01e60ba9849f
23436c6d-280d-481c-988c-7beead657e69

After optimization, it only collect 1000+ manifests and 60000+ data file metas, the procedure only cost 20min on local machine:
e46b828e-1013-4e81-86c2-b92e5daae241
c6ff7173-50f9-4e50-ba72-9d02bf712a20

File: paimon-core/src/main/java/org/apache/paimon/operation/LocalOrphanFilesClean.java [hunk 2]
```diff
@@ -114,7 +118,7 @@ private List<String> getUsedFiles(String branch) {
         ManifestFile manifestFile =
                 table.switchToBranch(branch).store().manifestFileFactory().create();
         try {
-            List<String> manifests = new ArrayList<>();
+            Set<String> manifests = new HashSet<>();
```
Comment (bknbkn) [inline paimon-core/src/main/java/org/apache/paimon/operation/LocalOrphanFilesClean.java:121]: While reading multiple snapshots, there will be duplicate manifests. Deduplication can improve efficiency.

**Review by JingsongLi — APPROVED** [review]
Good point! Thanks @bknbkn

**Pull request merged**